subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/codenarc/CodeNarcPluginToolchainsIntegrationTest.groovy

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.plugins.quality.codenarc
+
+import org.gradle.integtests.fixtures.AvailableJavaHomes
+import org.gradle.integtests.fixtures.MultiVersionIntegrationSpec
+import org.gradle.integtests.fixtures.TargetCoverage
+import org.gradle.integtests.fixtures.jvm.JavaToolchainFixture
+import org.gradle.internal.jvm.Jvm
+import org.gradle.quality.integtest.fixtures.CodeNarcCoverage
+import org.gradle.test.fixtures.file.TestFile
+
+/**
+ * Tests to ensure toolchains specified by the {@code CodeNarcPlugin} and
+ * {@code CodeNarc} tasks behave as expected.
+ */
+@TargetCoverage({ CodeNarcCoverage.getSupportedVersionsByJdk() })
+class CodeNarcPluginToolchainsIntegrationTest extends MultiVersionIntegrationSpec implements JavaToolchainFixture {
```

Comment:
Do we have a test that shows that we can analyze CodeNarc running with Groovy 4 while Gradle is using Groovy 3?

Reply:
Yeah, we added this (`CodeNarcGroovyVersionIntegrationTest`) when we backed out the change previously.  I've updated the test so that we test against both Groovy3 and Groovy4 (so that when we update to Gradle using Groovy4, we'll then be testing with codenarc using Groovy3).